Explorer/Assets/Scripts/CRDT/Memory/CRDTPooledMemoryAllocator.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Buffers;
+
+namespace CRDT.Memory
+{
+    /// <summary>
+    /// Uses ArrayPool under the hood
+    /// </summary>
+    public class CRDTPooledMemoryAllocator : ICRDTMemoryAllocator
+    {
+        private class MemoryOwner : IReadOnlyMemoryOwner<byte>
+        {
+            private readonly byte[] array;
+
+            private bool disposed;
+
+            internal MemoryOwner(byte[] array, int size)
+            {
+                this.array = array;
+                ReadOnlyMemory = this.array;
+                ReadOnlyMemory = ReadOnlyMemory.Slice(size);
+            }
+
+            public void Dispose()
+            {
+                if (!disposed)
+                {
+                    ArrayPool<byte>.Shared.Return(array);
+                    disposed = true;
+                }
+            }
+
+            public ReadOnlyMemory<byte> ReadOnlyMemory { get; }
+        }
+
+        public IReadOnlyMemoryOwner<byte> GetMemoryBuffer(in ReadOnlyMemory<byte> originalStream, int shift, int length)
+        {
+            var memoryOwner = MemoryPool<byte>.Shared.Rent(length);
+            var slice = originalStream.Slice(shift, length);
+            slice.CopyTo(memoryOwner.Memory);
+
+            return new MemoryOwner(ArrayPool<byte>.Shared.Rent(length), length);
```

Comment:
I am assuming that this class would not be used in a larger scale multi-threaded situation

Reply:
I think it is fine, I have read some warnings that used in a largely multithreaded environment the ArrayPool can fallback to always having to allocate memory when the default buffers are not large enough. I think for our cases it will be just fine.